pkg/target/http.go

```diff
@@ -399,55 +399,60 @@ func (ht *HTTPTarget) Write(messages []*models.Message) (*models.TargetWriteResu
 			}
 
 			// Ensure response body is always closed
-			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-				// Drain and close the body for successful responses
-				_, _ = io.Copy(io.Discard, resp.Body)
-				resp.Body.Close()
-
-				for _, msg := range goodMsgs {
-					if msg.AckFunc != nil { // Ack successful messages
-						msg.AckFunc()
+			func() {
```

Comment:
wrapping in function guarantees `resp.Body.Close()` is called after each iteration

Reply:
I see the reasoning here - I'm not confident that the loop is actually responsible for the problem. When reading from kinesis, the client we use provide data in batches of 1.

So, while you're correct that the defer function wouldn't get evaluated until after the loop, that _should_ actually happen immediately/very soon afterwards regardless. But this seems not to have been the case.

I have two theories about the actual explanation:

1. Not draining the body first causes the `Close()` method to fail, or wait for a long time expecting a closed body.
2. Something gnarly about deferred functions under high-concurrency doesn't operate according to my expectations

I'm not 100% certain - I'll need to dig into it and do some creative thinking about testing for those two theories (and/or find others), but once I do that I planned on considering changes to suit. In the meantime, your initial implementation should avoid both - so as long as it works well under load I'd be relatively confident about it even without a full explanation.

Ofc if it doesn't behave itself under load, or we see something that doesn't fir those theories there's further thinking to do.